crates/core/component/stake/src/component/action_handler/undelegate.rs

```diff
@@ -76,3 +44,83 @@ impl ActionHandler for Undelegate {
         Ok(())
     }
 }
+
+/// Interfaces for enforcing [`ActionHandler`] invariants.
+impl Undelegate {
+    /// Returns `Ok(())` if this delegation is from the current epoch.
+    ///
+    /// This ensures that the unbonding delay is enforced correctly. Additionally, it helps us
+    /// provide a more helpful error message if an epoch boundary was crossed.
+    async fn check_that_undelegate_was_prepared_for_current_epoch(
+        &self,
+        state: &impl StateWrite,
+    ) -> Result<()> {
```

Comment:
We should not structure the code this way.

It spreads the consensus critical part of the ActionHandler out into different places, rather than than having it in one place where it can be inspected locally.

And it leaks the internals of that implementation into a public interface, with methods that may or may not appear depending on compile settings, because the Undelegate type must be independent of the component implementation.

Reply:
https://github.com/penumbra-zone/penumbra/pull/4110#discussion_r1543180355

Reply:
this was reverted in 7bcd43c

Reply:
https://github.com/penumbra-zone/penumbra/pull/4110/commits/b513ddbcfc7bd44117eeffbb4815403b8306d108

i've reverted this in b513ddbcfc7bd44117eeffbb4815403b8306d108

Reply:
...i'm going to mark this as resolved, now that this change has been reverted.